src/github.com/matrix-org/dendrite/clientapi/routing/register.go

```diff
@@ -437,102 +271,72 @@ func handleRegistrationFlow(
 	accountDB *accounts.Database,
 	deviceDB *devices.Database,
 ) util.JSONResponse {
-	// TODO: Shared secret registration (create new user scripts)
-	// TODO: Enable registration config flag
-	// TODO: Guest account upgrading
 
-	// TODO: Handle loading of previous session parameters from database.
-	// TODO: Handle mapping registrationRequest parameters into session parameters
-
-	// TODO: email / msisdn auth types.
-
-	if cfg.Matrix.RegistrationDisabled && r.Auth.Type != authtypes.LoginTypeSharedSecret {
-		return util.MessageResponse(http.StatusForbidden, "Registration has been disabled")
+	if res := handleLoginTypeSharedSecret(req, r, cfg, sessionID); res != nil {
+		return *res
 	}
 
-	switch r.Auth.Type {
-	case authtypes.LoginTypeRecaptcha:
-		// Check given captcha response
-		resErr := validateRecaptcha(cfg, r.Auth.Response, req.RemoteAddr)
-		if resErr != nil {
-			return *resErr
-		}
-
-		// Add Recaptcha to the list of completed registration stages
-		sessions.AddCompletedStage(sessionID, authtypes.LoginTypeRecaptcha)
-
-	case authtypes.LoginTypeSharedSecret:
-		// Check shared secret against config
-		valid, err := isValidMacLogin(cfg, r.Username, r.Password, r.Admin, r.Auth.Mac)
-
-		if err != nil {
-			return httputil.LogThenError(req, err)
-		} else if !valid {
-			return util.MessageResponse(http.StatusForbidden, "HMAC incorrect")
-		}
-
-		// Add SharedSecret to the list of completed registration stages
-		sessions.AddCompletedStage(sessionID, authtypes.LoginTypeSharedSecret)
-
-	case authtypes.LoginTypeApplicationService:
-		// Check Application Service register user request is valid.
-		// The application service's ID is returned if so.
-		appserviceID, err := validateApplicationService(cfg, req, r.Username)
+	appservice, err := validateApplicationService(cfg, req)
```

Comment:
it's confusing here that `err` is not an `error`. Why not use `jsonRes` here and at line 275 (https://github.com/matrix-org/dendrite/pull/413/files#diff-789859953fb6c45e704d547fce33b9eeR275)?